tests/model/architecture.test.js

```diff
@@ -0,0 +1,190 @@
+
+var topLayer = new Layer('top-layer', 'unique-top-layer-id', [component1, component2], [connection1, connection2]);
+var botLayer = new Layer('bot-layer', 'unique-bot-layer-id', [component3, component4], [connection3]);
+
+describe('initialization', () => {
+    describe('constructor', () => {
+        test('parameters', () => {
+            let a = new Architecture('arch', [topLayer, botLayer], 5, 51);
+
+            expect(a.name).toBe('arch');
+            expect(a.layers).toEqual([topLayer, botLayer]);
+            expect(a.xSpan).toBe(5);
+            expect(a.ySpan).toBe(51);
+            expect(a.hasParams).toBe(false);
+        });
+
+        test('default', () => {
+            let a = new Architecture();
+
+            expect(a.name).toBe(Validation.DEFAULT_STR_VALUE);
+            expect(a.layers).toEqual([]);
+            expect(a.xSpan).toBe(Validation.DEFAULT_SPAN_VALUE);
+            expect(a.ySpan).toBe(Validation.DEFAULT_SPAN_VALUE);
+            expect(a.hasParams).toBe(false);
+        });
+    });
+
+    test('modify fields', () => {
+        let a = new Architecture();
+
+        a.name = 'arch';
+        a.layers = [topLayer, botLayer];
+        a.xSpan = 10;
+        a.ySpan = 15;
+        a.hasParams = true;
+
+        expect(a.name).toBe('arch');
+        expect(a.layers).toEqual([topLayer, botLayer]);
+        expect(a.xSpan).toBe(10);
+        expect(a.ySpan).toBe(15);
+        expect(a.hasParams).toBe(true);
+    });
+});
+
+describe('determineParams', () => {
+    test('does have params', () => {
+        let a = new Architecture();
+
+        a.xSpan = 10;
+        a.ySpan = 15;
+        expect(a.determineParams()).toBe(true);
+
+        expect(a.hasParams).toBe(true);
+    });
+
+    test('does not have params', () => {
+        let a = new Architecture();
+
+        a.xSpan = -14;
+        a.ySpan = -9876543;
+        expect(a.determineParams()).toBe(false);
+
+        expect(a.hasParams).toBe(false);
+    });
+});
+
+describe('validation', () => {
+    describe('valid', () => {
+        test('with params', () => {
+            let a = new Architecture('arch', [topLayer, botLayer], 1000, 1000);
+
+            expect(a.validate()).toBe(true);
+        });
+
+        test('without params', () => {
+            let a = new Architecture('arch', [topLayer, botLayer]);
+
+            expect(a.validate()).toBe(true);
+        });
+    });
+
+    describe('invalid', () => {
+        test('all fields', () => {
+            let a = new Architecture();
+
+            expect(a.validate()).toBe(false);
+        });
+
+        test('name', () => {
+            let a = new Architecture(Validation.DEFAULT_STR_VALUE, [topLayer, botLayer]);
+
+            expect(a.validate()).toBe(false);
+        });
+
+        describe('layers', () => {
+            test('empty', () => {
+                let a = new Architecture('arch', []);
+
+                expect(a.validate()).toBe(false);
```

Comment:
There is an interesting question here about if a more or less "empty" file should be valid, but we can come back to that later if necessary.

Reply:
Should we leave it as valid for now and just output an error message?

Reply:
I think its fine this way because we can use it as a "don't move forward with building an image" flag and then emit a specific message by checking for this particular case.